### PR TITLE
Fixing precision and trade limit issues

### PR DESCRIPTION
**Pull request by xeroquark**
- Added USD and coin precisions
- Fixed minimum balances to execute a trade

### CODE REVIEW

File: lib/env/TradingEnv.py [hunk 1]
```diff
@@ -69,18 +71,19 @@ def _take_action(self, action):
         cost_of_btc = 0
         revenue_from_sold = 0
 
-        if action == 0:
-            price = current_price * (1 + self.commission)
-            btc_bought = self.balance / price
-            cost_of_btc = self.balance
+        if action == 0 and self.balance >= 0.001:
```
Comment (georgeMuraveiAlkh) [inline lib/env/TradingEnv.py]: Every exchange and every token has different min order amount. IMO better to use additional parameters like `min_usd_order` and `min_token_order` to pass it in future from ccxt library for specific exchange.

Comment (xeroquark) [reply lib/env/TradingEnv.py]: Ah - yes it makes sense. I didn't know that ccxt provides this information. I'd have to look into it.

File: lib/env/TradingEnv.py [hunk 2]
```diff
@@ -20,8 +19,11 @@ def __init__(self, data_provider: BaseDataProvider, initial_balance=10000, commi
 
         self.logger = init_logger(__name__, show_debug=kwargs.get('show_debug', True))
 
+        self.usd_precision = 3
```
Comment (notadamking) [inline lib/env/TradingEnv.py]: Rename `usd_precision` -> `fiat_precision` and `coin_precision` -> `asset_precision`.

`fiat_precision` and `coin_precision` (`asset_precision`) should be allowed to be passed in as `kwargs`.

I.e. `self.fiat_precision = kwargs.get('fiat_precision', 3)`

File: lib/env/TradingEnv.py (hunk 1 above)
Comment (notadamking) [reply lib/env/TradingEnv.py]: Agreed, I do not like the hard-coded `0.001` here.

File: lib/env/TradingEnv.py [hunk 3]
```diff
@@ -69,18 +71,19 @@ def _take_action(self, action):
...
         revenue_from_sold = 0
 
-        if action == 0:
-            price = current_price * (1 + self.commission)
-            btc_bought = self.balance / price
-            cost_of_btc = self.balance
+        if action == 0 and self.balance >= 0.001:
+            price = round(current_price * (1 + self.commission), self.usd_precision)
+            btc_bought = round(self.balance / price, self.coin_precision)
+            cost_of_btc = round(self.balance, self.coin_precision)
```
Comment (notadamking) [inline lib/env/TradingEnv.py]: `self.balance` is in fiat (USD), so we should use `self.fiat_precision` here.

File: lib/env/TradingEnv.py (hunk 2 above)
Comment (xeroquark) [reply lib/env/TradingEnv.py]: Considering what you just said I think we should stick to the actual naming convention: base and quote (counter) currencies. 
So, I'd suggest `base_precision` and `quote_precision` instead.

**Comment by xeroquark** [comment]
Making the code more generic to handle different currency pairs (i.e., crypto/crypto & fiat/crypto, etc.) would require some changes in `TradingEnv.py` that will have a cascade effect in other parts of the code.

Currently, `data_provider` is of type `BaseDataProvider` class, which doesn't use `ccxt` and thus we cannot obtain precision values of the base and quote currency. Similarly, minimum (and maximum) price limits cannot be incorporated.  

I believe `ExchangeDataProvider` should become the base class or the code needs to be modified such that 
`def __init__(self, data_provider: BaseDataProvider, ....`
becomes
`def __init__(self, data_provider: ExchangeDataProvider, ....`

This would then require some further changes. @notadamking , what do you think?

**Comment by notadamking** [comment]
@xeroquark you're on the right track, but I think a better path would be to add the necessary interface methods to `BaseDataProvider`, and then implement them differently in `ExchangeDataProvider` and `StaticDataProvider`. For example, something like  `get_precision()`, `get_max_price()`, 
`get_min_price()` should do the trick.

**Comment by xeroquark** [comment]
Yeah - I was thinking of the same actually. I'll give it a try.

**Comment by notadamking** [comment]
Great! Let me know when it's ready to review.

**Comment by xeroquark** [comment]
@notadamking , the JSON structure of precision and min/max trading limits in `CCXT` package changes from exchange to exchange - based on the type of market (crypto-crypto or fiat-crypto etc.), which renders implementing this feature quite cumbersome.

Here are two JSON examples:
### Bittrex (ETH/BTC)
`...
'precision': {'amount': 8, 'price': 8},
 'limits': {'amount': {'min': 0.00740642, 'max': None},
  'price': {'min': 1e-08, 'max': None}},
...`

### Binance (ETH/BTC)
`...
 'precision': {'base': 8, 'quote': 8, 'amount': 3, 'price': 6},
 'limits': {'amount': {'min': 0.001, 'max': 100000.0},
  'price': {'min': 1e-06, 'max': 100000.0},
  'cost': {'min': 0.001, 'max': None}},
...`

Bittrex has some additional elements while it doesn't have min/max cost for a trade. In case of fiat-crypto markets and their JSON structures a similar mismatch can be also seen...

**Comment by notadamking** [comment]
@xeroquark the methodology for this library so far has been to provide the best defaults we can, but allow for extension wherever possible. I.e. by default implement Coinbase/Binance's min/max limits, and allow a `min_precision_key`, `max_precision_key`, `min_limits_key`, `max_limits_key` to be passed in as `kwargs`.

File: lib/env/TradingEnv.py (hunk 2 above)
Comment (sph3rex) [reply lib/env/TradingEnv.py]: I agree with the latest, because fiat will not always be present

**Comment by xeroquark** [comment]
@notadamking, PR is ready as far as I'm concerned. Fiat<->crypto pairs should work fine for almost all exchanges as I'm using CCXT's unified API. Coinbase Pro market values are used as default.

**Comment by notadamking** [comment]
Most of the suggestions in this PR have been implemented in the latest version of the repo, so I am going to close this for now.

**Pull request closed without merging**